File: frontend/src/types/forecast.ts
export interface ForecastPoint {
  period: string;
  historicalRevenue?: number | null;
  predictedRevenue: number | null;

  expectedOrders?: number | null;
  predictedOrders?: number | null;
  predictedAOV?: number | null;

  lowerBound: number | null;
  upperBound: number | null;
}

export interface ModelComparisonEntry {
  algorithm: string;
  wape: number | null;
  improvementOverBaselinePercent?: number | null;
  improvementOverV4Percent?: number | null;
}

export interface ModelComparison {
  baseline?: ModelComparisonEntry;
  v4?: ModelComparisonEntry;
  v4_1?: ModelComparisonEntry;
  bestModel?: string;
}

export interface ForecastTrend {
  direction:
    | "increasing"
    | "decreasing"
    | "stable"
    | "insufficient-data";

  changePercent: number | null;

  recentOrders?: number | null;
  projectedOrders?: number | null;

  recentAOV?: number | null;
  projectedAOV?: number | null;

  recentRevenue?: number | null;
  projectedRevenue?: number | null;
}

export interface RevenueOutlook {
  next7Days: number;
  next30Days: number;
  averageDailyRevenue: number;

  peakDay?: {
    date: string;
    predictedRevenue: number;
  } | null;

  lowestDay?: {
    date: string;
    predictedRevenue: number;
  } | null;
}

export interface PeakDay {
  date: string;
  predictedRevenue: number;
  rank?: number;
  deviationFromAveragePercent?: number | null;
}

export interface RiskAlert {
  level: "low" | "medium" | "high" | string;
  type: string;
  title?: string;
  message: string;
}

export interface Recommendation {
  priority: "low" | "medium" | "high" | string;
  category: string;

  title?: string;
  metric?: string;
  changePercent?: number | null;

  recommendation: string;

  supportingDates?: string[];
}

export interface BusinessInsights {
  generatedAt: string;

  revenueOutlook: RevenueOutlook;

  demandTrend: ForecastTrend;

  aovTrend: ForecastTrend;

  revenueTrend: ForecastTrend;

  peakDays: PeakDay[];

  riskAlerts: RiskAlert[];

  recommendations: Recommendation[];
}

export interface ForecastResponseData {
  algorithm: string;

  modelVersion: string;

  horizon: number;

  historyPoints: number;

  forecastCount: number;

  totalPredictedRevenue: number;

  mae: number;
  rmse: number;
  mape: number;
  wape: number;
  smape: number;

  evaluationDays: number;

  evaluation?: {
    status?: string;
    confidence?: string;
    explanation?: string;
    [key: string]: unknown;
  } | null;

  dataQuality?: {
    quality?: string;
    observedDays?: number;
    calendarDays?: number;
    [key: string]: unknown;
  } | null;

  modelDiagnostics?: Record<string, unknown> | null;

  baselineComparison?: Record<string, unknown> | null;

  modelComparison: ModelComparison | null;

  businessInsights: BusinessInsights | null;

  /**
   * Day-level explanations generated from the
   * existing V4.1 forecast output.
   *
   * This does not change the forecasting model.
   */
  forecastExplainability?:
    | ForecastExplanation[]
    | null;

  points: ForecastPoint[];

  generatedAt?: string;
}

export interface ForecastApiResponse {
  success: boolean;

  message?: string;

  data: ForecastResponseData;
}

export interface ForecastExplanationReason {
  type: string;
  message: string;
  impact: "positive" | "negative" | string;
}

export interface ForecastExplanation {
  date: string;
  weekday: string | null;

  predictedRevenue: number;
  expectedOrders: number;
  predictedAOV: number;

  revenuePosition:
    | "high"
    | "low"
    | "normal"
    | "unknown"
    | string;

  deviationFromAveragePercent:
    number | null;

  reasons: ForecastExplanationReason[];

  businessImplication: string;
}

File: ml-service/app/schemas/forecast.py
from typing import List, Optional

from pydantic import BaseModel, Field


class ForecastHistoryItem(BaseModel):
    date: str
    revenue: float = Field(ge=0)
    orders: int = Field(default=0, ge=0)
    observed: Optional[bool] = None


class ForecastRequest(BaseModel):
    history: List[ForecastHistoryItem]
    horizonDays: int = Field(default=30, ge=1, le=365)


class ForecastPoint(BaseModel):
    date: str
    predictedRevenue: float

    # Continuous statistical expectation.
    # Unlike predictedOrders, this is not rounded.
    expectedOrders: Optional[float] = None

    # Whole-number transaction estimate shown to users.
    predictedOrders: Optional[int] = None

    predictedAOV: Optional[float] = None
    lowerBound: Optional[float] = None
    upperBound: Optional[float] = None


class ForecastDataQuality(BaseModel):
    quality: str
    observedDays: int
    calendarDays: int
    missingDays: int


class ForecastEvaluationDay(BaseModel):
    date: str
    weekday: str
    actualRevenue: float
    v4Prediction: float
    v4_1Prediction: float
    baselinePrediction: float
    v4AbsoluteError: float
    v4_1AbsoluteError: float
    baselineAbsoluteError: float
    v4ErrorPercent: Optional[float] = None
    v4_1ErrorPercent: Optional[float] = None
    baselineErrorPercent: Optional[float] = None
    orders: int
    aov: Optional[float] = None
    observed: Optional[bool] = None


class ForecastEvaluation(BaseModel):
    status: str
    confidence: str
    evaluationDays: int
    minimumRecommendedDays: int
    message: Optional[str] = None
    evaluationByDay: List[ForecastEvaluationDay] = Field(default_factory=list)


class ForecastResponse(BaseModel):
    success: bool
    algorithm: str
    modelVersion: str

    mae: Optional[float] = None
    rmse: Optional[float] = None
    mape: Optional[float] = None

    # Sparse-data-aware evaluation metrics
    wape: Optional[float] = None
    smape: Optional[float] = None
    evaluationDays: Optional[int] = None

    # Evaluation reliability/status
    evaluation: Optional[ForecastEvaluation] = None

    # Historical data quality
    dataQuality: Optional[dict] = None
    modelDiagnostics: Optional[dict] = None
    baselineComparison: Optional[dict] = None
    modelComparison: Optional[dict] = None
    expectedOrders: Optional[float] = None

    forecast: List[ForecastPoint]